### DISCUSSION TITLE
New release planned?

### DISCUSSION THREAD

**Discussion by sid-6581** [morelinq/MoreLINQ#798] [General]
It's been a while since the last release, and some useful things like nullable support seems to have already been added but not published. Is there a new release planned, or any way to get alpha/beta nuget packages?

**Comment by atifaziz** [comment]
A release has been overdue since quite some time now and while [the 3.4.0 milestone](https://github.com/morelinq/MoreLINQ/milestone/24) has no open items, I did want to give the nullability annotations another review in light of recent developments. Lacking the time to do that is probably what's keeping me from getting it out the door. While I can't commit to any deadline…

> or any way to get alpha/beta nuget packages?

CI builds are published at: https://www.myget.org/F/morelinq/api/v3/index.json

**Reply by sid-6581** [reply]
Thank you! If there is anything I could help review I would be happy to do so. I've used MoreLINQ for quite some time, but haven't contributed, so I'm not quite sure where things stand and what needs to be done.

**Reply by atifaziz** [reply]
Hey, appreciate your offer to help! I just posted a fix for a bug (see PR #800) that was introduced with the nullable reference annotations work done in PR #582. I've also submitted PR #801 as follow-up to a to-do that was pending from PR #582. I'd appreciate your review as another pair of eyes can always help.

As I said earlier, I feel that all the annotations work in PR #582 needs a review in light of changes with C# 9. If I spawn new PRs then I'll tag you directly for review, but if you'd like to help with speeding that up too then feel free to submit PRs where you can spot improvements. As a quick starting point, I'd like to see if the uses of [null-forgiving operator (`!`)](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/null-forgiving) that were required previously are still necessary. For example, in PR #801, they all go away for one of the methods!

**Reply by sid-6581** [reply]
Great! I will take a look over the next couple of days. I've had to put a lot of work into converting my own projects to use nullable reference types, so I'll see if I can catch anything!

**Reply by sid-6581** [reply]
I looked at #800 and #801 last night, and those both look good to me. I started looking at the rest of the codebase as well, and saw a couple of things I want to look at closer today. One of the first things I noticed is that there are parameters that are treated as nullable, but aren't marked as such. Here's one example (`comparer`):

https://github.com/morelinq/MoreLINQ/blob/dae16605c05a46eaae32ce59e325a19352279fd8/MoreLinq/FullGroupJoin.cs#L142-L150

There are other instances like this one,, as well as other similar issues. I assume this is an example of something that should be cleaned up? If so, I can hunt these down and try to clean them up in a PR. Just wanted to make sure I'm on the right track, because in one of the locations there was a TODO saying nulls would be disallowed in the next major release. I don't want to change public APIs to nullable if the intent is that null support is going to be dropped.

**Reply by atifaziz** [reply]
That's a really good catch and seems to have been an oversight that slipped out of PR #582.

> There are other instances like this one,, as well as other similar issues.

I ran a quick and approximate grep for just `IEqualityComparer`:

    git grep IEqualityComparer MoreLinq | grep -v /// | grep -v -F ?

and you're right:

    MoreLinq/Collections/Dictionary.cs:        public Dictionary(IEqualityComparer<TKey> comparer)
    MoreLinq/CountBy.cs:                void Loop(IEqualityComparer<TKey> cmp)
    MoreLinq/Extensions.g.cs:            IEqualityComparer<TKey> comparer)
    MoreLinq/Extensions.g.cs:            IEqualityComparer<TKey> comparer)
    MoreLinq/Extensions.g.cs:            TSource separator, IEqualityComparer<TSource> comparer,
    MoreLinq/FullGroupJoin.cs:            IEqualityComparer<TKey> comparer)
    MoreLinq/FullGroupJoin.cs:            IEqualityComparer<TKey> comparer)
    MoreLinq/FullGroupJoin.cs:            IEnumerable<TResult> _(IEqualityComparer<TKey> comparer)
    MoreLinq/GroupAdjacent.cs:            IEqualityComparer<TKey> comparer)
    MoreLinq/Lookup.cs:        private IEqualityComparer<TKey> _comparer;
    MoreLinq/Lookup.cs:        internal static Lookup<TKey, TElement> Create<TSource>(IEnumerable<TSource> source, Func<TSource, TKey> keySelector, Func<TSource, TElement> elementSelector, IEqualityComparer<TKey> comparer)
    MoreLinq/Lookup.cs:        internal static Lookup<TKey, TElement> CreateForJoin(IEnumerable<TElement> source, Func<TElement, TKey> keySelector, IEqualityComparer<TKey> comparer)
    MoreLinq/Lookup.cs:        private Lookup(IEqualityComparer<TKey> comparer)
    MoreLinq/RunLengthEncode.cs:            IEnumerable<KeyValuePair<T, int>> _(IEqualityComparer<T> comparer)
    MoreLinq/ScanBy.cs:            IEnumerable<KeyValuePair<TKey, TState>> _(IEqualityComparer<TKey> comparer)
    MoreLinq/Split.cs:            TSource separator, IEqualityComparer<TSource> comparer,

> If so, I can hunt these down and try to clean them up in a PR. Just wanted to make sure I'm on the right track, …

While not all might apply, you seem to be on the right track 👍🏼 (especially for the public surface) so a PR would most certainly be welcome.

> …one of the locations there was a TODO saying nulls would be disallowed in the next major release.

I opened issue #802 to track one of those earlier. Is that the one? Did you comes across others?

> I don't want to change public APIs to nullable if the intent is that null support is going to be dropped.

Thanks for checking first!

By the way, in PR #804 for `Pad`, I extended the tests to reference types and enabled nullable context. I am not very happy with the test duplication for sake of testing nullability at compilation time, but I haven't spent much time think about how to validate it otherwise. It seems it needs another class of tests with explicitly declared types (instead of using `var`). If you have any ideas or if you've across something similar in other projects then fire away!

**Reply by sid-6581** [reply]
#802 was the one I was thinking of, yes. I haven't come across any others like that yet, but I'm going through things one at a time.

I've submitted a couple of PRs that I think will improve things. There's still more PRs to come, but we'll get there! For the most part things look good as far as I can tell.

I'll also check out the `Pad` tests to see, I haven't had a chance to look at that yet.

**Reply by Arithmomaniac** [reply]
> CI builds are published at: https://www.myget.org/F/morelinq/api/v3/index.json

Any chance we could get one of them published as a beta? There's two years of bug fixes waiting. (To be fair, there aren't a lot of bugs - kudos.)

**Reply by atifaziz** [reply]
I'll see if I can push out a beta in the coming days.

**Reply by frozenprakash** [reply]
Damn MoreLinq is seriously undermined!
I watched every single video on YouTube too and started using MoreLinq way too vastly in all my projects!!

Really awesome work!

![image](https://user-images.githubusercontent.com/16418079/135555297-054035ec-f054-4de9-88bb-b00a3455074f.png)

**Comment by atifaziz** [comment]
Life happens and so the latest release took a back seat for too long, but for what it's worth, [3.4 is now released](https://github.com/morelinq/MoreLINQ/tree/v3.4.0)! Thank you for your understanding.
